add Python module

Pokisiekk opened the pull request:
Signed-off-by: Kamil Pokornicki <kamil.pokornicki@3mdeb.com>

Commit:
add Python Module with one working function

Signed-off-by: Kamil Pokornicki <kamil.pokornicki@3mdeb.com>

Commit:
lib/TestModule.py: update function to work with one request

Signed-off-by: Kamil Pokornicki <kamil.pokornicki@3mdeb.com>

arturkow2000 commented on lib/TestModule.py:
```diff
@@ -0,0 +1,70 @@
+
+        with open(output_file, 'wb') as out:
+            out.write(encoded)
+            out.close()
```
There is no need to save output to a file, cbor should be returned to robotframework as object.

Pokisiekk replied:
applied

arturkow2000 commented on lib/TestModule.py:
```diff
@@ -0,0 +1,70 @@
+
+        # hexdump.hexdump(encoded)
+
+    async def send_data(self, endpoint, argument):
```
We have `send_data` function but we are not using it. We should this and completely remove usage of `coap-client`

Pokisiekk replied:
applied

arturkow2000 commented on lib/TestModule.py:
```diff
@@ -0,0 +1,70 @@
+        # hexdump.hexdump(encoded)
+
+    async def send_data(self, endpoint, argument):
+        proto = await Context.create_client_context()
```
Context should be created only once in library constructor - creating a new context on each request will cause problems later on as Fobnail will treat each request as coming from different client.

Pokisiekk replied:
I add commented code in `TestModule.py` - After adding the constructor, the whole module wasn't working. After a little research, I think `async` +` __init__` won't work together. With two calls to the `send_data` function, no problem appears, but as I understand that it can appears in the future.

arturkow2000 replied:
@Pokisiekk async functions returns coroutines which must be awaited before returning response, so the call to constructor didn't return `TestModule` object but something else, since constructors all handled by robotframework probably we won't be able to then (at least not async). In that case, context may during the first call to `send_data()` and kept for later usage.

`send_data` function must also be awaited, currently no data is sent to Fobnail at all:
```
../home/artur/.local/lib/python3.8/site-packages/robot/running/bodyrunner.py:50: RuntimeWarning: coroutine 'TestModule.send_data' was never awaited
  step.run(self._context, self._run, self._templated)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

There is library (https://github.com/Chetic/robotframework-async) which can handle `await` from robotframework, maybe you could use that. If not, then you can have an sync and async function, where sync functions calls async and blocks till async completes.

Pokisiekk replied:
I tried to use recommended library, but it didn't work correctly - I got the same problem: https://github.com/Chetic/robotframework-async/issues/5 
I used workaround from this issue: https://github.com/robotframework/robotframework/issues/4089

arturkow2000 replied:
@Pokisiekk I made a few improvements - see my commits

arturkow2000 commented on lib/TestModule.py:
```diff
@@ -0,0 +1,70 @@
+        proto = await Context.create_client_context()
+
+        if requests[endpoint]['in'] == 'cbor':
+            payload = open(argument, 'rb').read()
```
We should not read data from files, data should be passed as an object

Pokisiekk replied:
applied

arturkow2000 commented on lib/TestModule.py:
```diff
@@ -0,0 +1,70 @@
+            if r.code != Code.CREATED:
+                raise RuntimeError(f'Request failed with error {r.code}')
+            if requests[endpoint]['out'] == 'csr':
+                with open('provisioning/tmp/fobnail.csr', 'w+b') as out:
```
Same as above, that should be returned as an object instead of writing it file.

Pokisiekk replied:
I can't use returned object in openssl commands, so I left it as it is for now. You can check the test-log on the cloud for more details.

arturkow2000 replied:
@Pokisiekk It shoud be possible to spawn openssl command, write csr to stdin and retrieve certificate from stdout. But let's leave it as is, for now.

arturkow2000 commented:
`Generate certificates` keyword could accept CSR as parameter and return certificate on success. After running `openssl` command we can write CSR to stdin and read the certificate from stdout.

sulewskiprzemyslaw commented on lib/TestModule.py:
```diff
@@ -0,0 +1,70 @@
+from cryptography import x509
```
If we create RF library, we should name it in accordance with the performed operations. So, this module should be named `Token provisioning` or something in that way.

Pokisiekk replied:
RF library already performs two quite different functions and will definitely be developed in the future. This file is something like `keywords.robot` and the name should also be quite universal. I know that `TestModule` is pretty general, but I can't think of anything better.

sulewskiprzemyslaw replied:
OK, on future tasks, let's just try to think it through and plan it out .

Commit:
keywords.robot: Generate Certificates: add commented code

Signed-off-by: Kamil Pokornicki <kamil.pokornicki@3mdeb.com>

Commit:
tests/token_provision.robot: add support for  Send Data from TestModule, some updates

Signed-off-by: Kamil Pokornicki <kamil.pokornicki@3mdeb.com>

Commit:
keywords.robot: remove comments

Signed-off-by: Kamil Pokornicki <kamil.pokornicki@3mdeb.com>

Commit:
lib/TestModule.py: update send data function

Signed-off-by: Kamil Pokornicki <kamil.pokornicki@3mdeb.com>

arturkow2000 committed:
lib/TestModule.py: fix error handling

Don't catch NetworkError, don't swapn new thread just for asyncio, doing
so hides Python exceptions, giving false-positive test results.

arturkow2000 committed:
keywords.robot: fix certificate generation
